Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -1263,4 +1275,237 @@ lemma nonsingularLift_addMap {P Q : PointClass F} (hP : W.NonsingularLift P)
+  rfl
+
+/-- The map from a nonsingular rational point on a Weierstrass curve `W'` in affine coordinates
+to the corresponding nonsingular rational point on `W'` in Jacobian coordinates. -/
+def fromAffine [Nontrivial R] : W'.toAffine.Point → W'.Point
+  | 0 => 0
+  | Affine.Point.some h => ⟨(nonsingularLift_some ..).mpr h⟩
+
+lemma fromAffine_zero [Nontrivial R] : fromAffine 0 = (0 : W'.Point) :=
+  rfl
+
+lemma fromAffine_some [Nontrivial R] {X Y : R} (h : W'.toAffine.Nonsingular X Y) :
+    fromAffine (Affine.Point.some h) = ⟨(nonsingularLift_some ..).mpr h⟩ :=
+  rfl
+
+/-- The negation of a nonsingular rational point on `W`.
+Given a nonsingular rational point `P` on `W`, use `-P` instead of `neg P`. -/
+def neg (P : W.Point) : W.Point :=
+  ⟨nonsingularLift_negMap P.nonsingular⟩
+
+instance instNegPoint : Neg W.Point :=
+  ⟨neg⟩
+
+lemma neg_def (P : W.Point) : -P = P.neg :=
+  rfl
+
+lemma neg_point (P : W.Point) : (-P).point = W.negMap P.point :=
+  rfl
+
+/-- The addition of two nonsingular rational points on `W`.
+Given two nonsingular rational points `P` and `Q` on `W`, use `P + Q` instead of `add P Q`. -/
+noncomputable def add (P Q : W.Point) : W.Point :=
+  ⟨nonsingularLift_addMap P.nonsingular Q.nonsingular⟩
+
+noncomputable instance instAddPoint : Add W.Point :=
+  ⟨add⟩
+
+lemma add_def (P Q : W.Point) : P + Q = P.add Q :=
+  rfl
+
+lemma add_point (P Q : W.Point) : (P + Q).point = W.addMap P.point Q.point :=
+  rfl
+
+end Point
+
+end Point
+
+section Affine
+
+/-! ### Equivalence with affine coordinates -/
+
+open scoped Classical
+
+namespace Point
+
+variable (W) in
+/-- The map from a point representative that is nonsingular on a Weierstrass curve `W` in Jacobian
+coordinates to the corresponding nonsingular rational point on `W` in affine coordinates. -/
+noncomputable def toAffine (P : Fin 3 → F) : W.toAffine.Point :=
+  if hP : W.Nonsingular P then if hPz : P z = 0 then 0 else
+    Affine.Point.some <| (nonsingular_of_Z_ne_zero hPz).mp hP else 0
+
+lemma toAffine_of_singular {P : Fin 3 → F} (hP : ¬W.Nonsingular P) : toAffine W P = 0 := by
+  rw [toAffine, dif_neg hP]
+
+lemma toAffine_of_Z_eq_zero {P : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z = 0) :
+    toAffine W P = 0 := by
+  rw [toAffine, dif_pos hP, dif_pos hPz]
+
+lemma toAffine_zero : toAffine W ![1, 1, 0] = 0 :=
+  toAffine_of_Z_eq_zero nonsingular_zero rfl
+
+lemma toAffine_of_Z_ne_zero {P : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z ≠ 0) :
+    toAffine W P = Affine.Point.some ((nonsingular_of_Z_ne_zero hPz).mp hP) := by
+  rw [toAffine, dif_pos hP, dif_neg hPz]
+
+lemma toAffine_some {X Y : F} (h : W.Nonsingular ![X, Y, 1]) :
+    toAffine W ![X, Y, 1] = Affine.Point.some ((nonsingular_some ..).mp h) := by
+  simp only [toAffine_of_Z_ne_zero h one_ne_zero, fin3_def_ext, one_pow, div_one]
+
+lemma toAffine_smul (P : Fin 3 → F) {u : F} (hu : IsUnit u) :
+    toAffine W (u • P) = toAffine W P := by
```

Comment:
Since `F` is a field, isn't it more conventional to assume `u ≠ 0`?

Reply:
Definitely, but this is consistent with the rest of the ring lemmas assuming `IsUnit`. I can add a version assuming `u ≠ 0` but some of the places where it's used are a bit more annoying to write with `u ≠ 0` (e.g. I could be writing a chain of dot notations `.neg`, `.sub`, `.pow`, `.div` instead of `neg_ne_zero`, `sub_ne_zero`, `pow_ne_zero`, `div_ne_zero`).

Reply:
Okay. Indeed it seems a big change to replace the condition.